### PR TITLE
Add FP16 datatype support for OIP grpc

### LINKED ISSUE #3643: Inference gRPC/Rest client to support FP16
/kind feature

**Describe the solution you'd like**
Triton client supports FP16, kserve rest/grpc should also support transporting FP16.
https://github.com/kserve/kserve/blob/master/python/kserve/kserve/protocol/infer_type.py#L278-L296

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Links to the design documents:**
[Optional, start with the short-form [RFC template](https://docs.google.com/document/d/1UcBeLfZ_JMGpVrPJmYtEIVH_9Y4U3AEKQdq_IKuOMrU) to outline your ideas and get early feedback.]
[Required, use the longer-form [design doc template](https://docs.google.com/document/d/1Mtoui_PP2a9N59NjYHnsvrdJ8t2iKFwIJAx1zRO_I1c) to specify and discuss your design in more detail]

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3643

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add FP16 datatype support for OIP grpc
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: python/kserve/kserve/protocol/infer_type.py [hunk 1]
```diff
@@ -562,7 +647,14 @@ def as_numpy(self) -> numpy.ndarray:
         if dtype is None:
             raise InvalidInput("invalid datatype in the input")
         if self._raw_data is not None:
-            np_array = np.frombuffer(self._raw_data, dtype=dtype)
+            if self.datatype == "BYTES":
+                # String results contain a 4-byte string length
+                # followed by the actual string characters. Hence,
+                # need to decode the raw bytes to convert into
+                # array elements.
+                np_array = deserialize_bytes_tensor(self._raw_data)
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/infer_type.py:652]: This is only for the REST case right? for gRPC we can just call np.frombuffer

File: python/kserve/kserve/protocol/infer_type.py [hunk 2]
```diff
@@ -32,43 +31,85 @@
 )
 from ..utils.numpy_codec import to_np_dtype, from_np_dtype
 
+# Dictionary that maps model names to a boolean indicating whether their output uses the FP16 datatype
+_model_name_to_fp16_map: Dict[str, bool] = {}
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/infer_type.py]: Why this is a global map?

**Comment by yuzisun** [comment]
Let's add an e2e test for this

File: python/kserve/kserve/protocol/infer_type.py (hunk 2 above)
Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py]: removed

File: python/kserve/kserve/protocol/infer_type.py (hunk 1 above)
Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py:652]: Triton client uses serialization for both http and grpc

File: python/kserve/kserve/protocol/grpc/servicer.py [hunk 3]
```diff
@@ -34,6 +36,21 @@ def __init__(
...
+            raise InvalidInput(
+                f"the number of inputs ({len(request.inputs)}) does not match the expected number of "
+                f"raw input contents ({raw_inputs_length}) for model '{request.model_name}'."
+            )
+        if raw_inputs_length != 0:
+            for input_ in request.inputs:
+                if input_.HasField("contents"):
+                    raise InvalidInput(
+                        f"contents field must not be specified when using raw_input_contents for input '{input_.name}' for model '{request.model_name}'"
+                    )
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/grpc/servicer.py:52]: I think only FP16 and BF16 requires setting raw inputs, for other types user do not have to set raw input contents

Comment (sivanantha321) [reply python/kserve/kserve/protocol/grpc/servicer.py:52]: If raw_inputs_contents specified, then the contents field should be empty. https://github.com/kserve/open-inference-protocol/blob/52528cf4c48368114874e99217d25b9e006e494a/specification/protocol/open_inference_grpc.proto#L203-L204

File: python/kserve/kserve/protocol/infer_type.py [hunk 4]
```diff
@@ -173,7 +211,14 @@ def as_numpy(self) -> np.ndarray:
         if dtype is None:
             raise InvalidInput(f"invalid datatype {dtype} in the input")
         if self._raw_data is not None:
-            np_array = np.frombuffer(self._raw_data, dtype=dtype)
+            if self.datatype == "BYTES":
+                # String results contain a 4-byte string length
+                # followed by the actual string characters. Hence,
+                # need to decode the raw bytes to convert into
+                # array elements.
+                np_array = deserialize_bytes_tensor(self._raw_data)
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/infer_type.py:219]: How `deserialize_bytes_tensor` is different from `np.frombuffer` ?

Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py:219]: Usually strings are serialized as a 1-dimensional array containing the 4-byte byte size followed by the actual element bytes. All elements are concatenated together. To deserialize this format we are using deserialize_byte_tensor

File: python/kserve/kserve/protocol/infer_type.py [hunk 5]
```diff
@@ -791,8 +880,14 @@ def to_grpc(self) -> ModelInferResponse:
         """
         infer_outputs = []
         raw_output_contents = []
+        use_raw_outputs = False
+        # If FP16 datatype is present in the outputs use raw outputs.
+        if _contains_fp16_datatype(self):
+            use_raw_outputs = True
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/infer_type.py:886]: What happens if only one of the outputs use FP16?

File: python/kserve/test/test_grpc_server.py [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+        self.name = name
+        self.ready = False
+
+    def load(self):
+        self.ready = True
+
+    async def predict(self, request, headers=None):
+        outputs = pd.DataFrame(
+            {
+                "fp16_output": request.get_input_by_name("fp32_input")
```
Comment (yuzisun) [inline python/kserve/test/test_grpc_server.py:81]: fp16_input

File: python/kserve/test/test_grpc_server.py [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+    }
+
+
+@pytest.mark.asyncio
+@patch(
+    "kserve.protocol.grpc.servicer.to_headers", return_value=[]
+)  # To avoid NotImplementedError from trailing_metadata function
+async def test_grpc_raw_inputs(mock_to_headers, server):
+    """
+    If we receive raw inputs then, the response also should be in raw output format.
```
Comment (yuzisun) [inline python/kserve/test/test_grpc_server.py:302]: Is this the right assumption ? wondering how triton behaves.

File: python/kserve/kserve/protocol/infer_type.py (hunk 5 above)
Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py:886]: The response will be returned as raw outputs

File: python/kserve/kserve/protocol/infer_type.py (hunk 4 above)
Comment (yuzisun) [reply python/kserve/kserve/protocol/infer_type.py:219]: what happens if user set the string dtype np.array bytes to the raw content field?

File: python/kserve/test/test_grpc_server.py (hunk 7 above)
Comment (sivanantha321) [reply python/kserve/test/test_grpc_server.py:302]: Yes, I just followed the triton behaviour

File: python/kserve/kserve/protocol/infer_type.py (hunk 5 above)
Comment (yuzisun) [reply python/kserve/kserve/protocol/infer_type.py:886]: same for output according to 
https://github.com/kserve/open-inference-protocol/blob/52528cf4c48368114874e99217d25b9e006e494a/specification/protocol/open_inference_grpc.proto#L257

Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py:886]: I mean the all the outputs will be sent as bytes which will be stored in raw_ouput_contents field. The contents field will be empty.

File: python/kserve/test/test_grpc_server.py [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+    """
+    raw_input_contents = [
+        np.array([6.8, 2.8, 4.8, 1.4, 6.0, 3.4, 4.5, 1.6], dtype=np.float32).tobytes(),
+        np.array([6, 2, 4, 1, 6, 3, 4, 1], dtype=np.int32).tobytes(),
+        serialize_byte_tensor(
```
Comment (yuzisun) [inline python/kserve/test/test_grpc_server.py]: Wondering if this is the same as `ndarray.tobytes`
https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tobytes.html

File: python/kserve/kserve/protocol/infer_type.py (hunk 4 above)
Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py:219]: serialize_bytes_tensor only used if the strings is sent in raw_input_contents. If we use contents we can just directly provide strings as list of bytes (e.g. [b"cat", b"dog"] )

File: python/kserve/test/test_grpc_server.py [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+                "parameters": {"binary_data_size": {"int64_param": "64"}},
+            },
+            {
+                "name": "fp16_output",
+                "datatype": "FP16",
+                "shape": ["8"],
+                "parameters": {"binary_data_size": {"int64_param": "16"}},
+            },
+        ],
+        "raw_output_contents": [
```
Comment (yuzisun) [inline python/kserve/test/test_grpc_server.py:549]: Can we add the asserts to decode raw output content to InferResponse/InferOutput so deserializeBytesTensor is tested?

File: python/kserve/kserve/protocol/infer_type.py [hunk 10]
```diff
@@ -791,8 +880,14 @@ def to_grpc(self) -> ModelInferResponse:
...
         for infer_output in self.outputs:
-            if isinstance(infer_output.data, numpy.ndarray):
+            if use_raw_outputs and infer_output.data:
+                infer_output.data = infer_output.as_numpy()
+            if isinstance(infer_output.data, np.ndarray):
                 infer_output.set_data_from_numpy(infer_output.data, binary_data=True)
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/infer_type.py]: seems like we are decoding back to numpy on line 889 and then serializing to bytes again on line 891. If it is already in bytes then we can simply pass them? For simple data it is fine but might be an overhead for large output tensor

File: python/kserve/test/test_grpc_server.py (hunk 9 above)
Comment (sivanantha321) [reply python/kserve/test/test_grpc_server.py:549]: Added

File: python/kserve/kserve/protocol/infer_type.py (hunk 10 above)
Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py]: This case is for list data. Added type check for list

**Pull request merged**